common/tasks/multi_destination_scheduler.go

```diff
@@ -0,0 +1,241 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tasks
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+)
+
+type RateLimiter interface {
+	Wait(context.Context) error
+}
+
+// LimiterOptions are options for a single destination's limiter.
+type LimiterOptions struct {
+	// Concurrency returns the concurrency limit.
+	Concurrency func() int
+	RateLimiter RateLimiter
+}
+
+// limiter acts a concurrency and rate limiter for a single destination.
+// All access to the limiter must be locked.
+type limiter[T Task] struct {
```

Comment:
I actually think the responsibilities of the `limiter` are a bit overloaded. I think we can and should split it up because part of it is a generic worker pool with some nice optimizations: it spins up goroutines on demand and leaves them open while there's still work immediately available. I'd extract out another type for this, and then put all the circuit breaking and rate limiting logic in the task function itself.

Here's a gist that shows that I had in mind: https://gist.github.com/MichaelSnowden/6b444e9fda887fa0227848a11290842c
We definitely don't have to use this, but I want to paint a clear image to show that there is quite a bit of logic here that should be extracted because:

1) It's complicated enough on its own
2) It has generic usefulness